server/socketManager.js

```diff
@@ -0,0 +1,373 @@
+const assert = require('bsert');
+const { sha256, random, ccmp } = require('bcrypto');
+const { base58 } = require('bstring');
+const { URL } = require('url');
+const Validator = require('bval');
+
+class SocketManager extends Server {
+  /**
+   * Create an http server.
+   * @constructor
+   * @param {Object} options
+   */
+  constructor(options) {
+    super(new SocketManagerOptions(options));
+
+    this.logger =
+      this.options.logger && this.options.logger.context('socket-manager');
+    this.clients = new Map();
+    this.types = ['node', 'wallet'];
+    this.init();
+  }
+
+  /*
+   * Initialize server
+   */
+  init() {
+    this.on('request', req => {
+      if (req.method === 'POST' && req.pathname === '/') return;
+
+      this.logger.debug(
+        'Request for method=%s path=%s (%s).',
+        req.method,
+        req.pathname,
+        req.socket.remoteAddress
+      );
+    });
+
+    this.on('listening', address => {
+      this.logger.info(
+        'SocketManager server listening on %s (port=%d).',
+        address.address,
+        address.port
+      );
+    });
+  }
+
+  /*
+   * handle new socket clients that connect to server
+   * @param {WebSocket} socket
+   * @returns {null}
+   */
+  handleSocket(socket) {
+    socket.hook('auth', (...args) => {
+      if (socket.channel('auth')) throw new Error('Already authed.');
+
+      if (!this.options.noAuth) {
+        const valid = new Validator(args);
+        const key = valid.str(0, '');
+        if (key.length > 255) throw new Error('Invalid API key.');
+
+        const data = Buffer.from(key, 'utf8');
+        const hash = sha256.digest(data);
+
+        if (!ccmp(hash, this.options.apiHash))
+          throw new Error('Invalid API key.');
+      }
+
+      socket.join('auth');
+      this.logger.info('Successful auth from %s.', socket.host);
+      this.handleAuth(socket);
+      return null;
+    });
+  }
+
+  /*
+   * Utility to verify that the pathname
+   * matches with the id of a client on the class
+   * @private
+   * @param {string} pathname - pathname to validate
+   * @returns {string} id
+   */
+  getIdFromPath(pathname) {
+    assert(
+      pathname[0] === '/' && pathname[pathname.length - 1] === '/',
+      'Invalid pathname'
+    );
+    const id = pathname.slice(1, pathname.length - 1);
+    if (!this.clients.has(id))
+      this.logger.warning(
+        'getIdFromPath: id %s from pathname %s does not exist',
+        id,
+        pathname
+      );
+
+    return id;
+  }
+
+  /*
+   * Handle new auth'd websocket.
+   * @private
+   * @param {WebSocket} socket
+   */
+  handleAuth(socket) {
+    this.logger.info('Preparing socket from %s', socket.url);
+    const parseEvent = event => {
+      assert(typeof event === 'string', 'Must pass a string to getClient');
+      // defaulting to node client if there is no client prefix
+      let parsedEvent = event;
+      let id = 'node';
+
+      // check if first full word is a valid type
```

Comment:
What defines a valid type?

Is it one of the socket emit types, like as seen here? -> http://bcoin.io/api-docs/index.html#using-wallet-sockets